### PR TITLE
TeamCity-HockeyApp-Kobiton

### PR DESCRIPTION
**Pull request by phanthaoIT**

### CODE REVIEW

File: run-automation-test.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+This part will guide you to how to configure TeamCity to integrate with your automation test GitHub repository. Skip this step if you have already had TeamCity setup.
+ + For instruction on how to install and configure TeamCity server, follow [this guide](https://confluence.jetbrains.com/display/TCD18/Installing+and+Configuring+the+TeamCity+Server).
+
+Set up your project and link your app repository that we will test with Kobiton later.
++ You need to configure a connection to your GitHub repository. Instruction on how to connect GitHub with TeamCity can be found [this guide](https://confluence.jetbrains.com/display/TCD10/Integrating+TeamCity+with+VCS+Hosting+Services), in `Connecting to GitHub` section. After the connection is configured, TeamCity will be granted with `full control` of private repositories and `Write repository hooks` permission.
+ 
+>More information about TeamCity can be found in [this documentation](https://confluence.jetbrains.com/display/TCD10/TeamCity+Documentation).
+## 2. Configure HockeyApp 
+HockeyApp is a service that allows developers to recruit and manage testers, distribute apps, and collect crash reports, among other things.
```
Comment (thangvo38) [inline run-automation-test.md]: This should be in `README.md`

File: run-automation-test.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+- **Get API Tokens**
+
+  In the top right corner, click **user icon** -> **My Devices** -> **API Tokens**.
+
+  Create a Hockey App API token for your app with `Full Access` or `Upload and Release` right.
+
+  ![Create App](./assets/img/2-token.png)
+### 2.2. Fetching the app download URL
+
+To get the app download URL, we have provided a script written in NodeJS. After executing that script, you will get download URL of that app, which will be used in the next step.
```
Comment (thangvo38) [inline run-automation-test.md]: Please be more specific. The script is used to get the `latest` app version.

File: run-automation-test.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+The desired capabilities need to be added to the automation test script to allow the test to be executed on Kobiton device.
+  + Click **Devices** in the top navigation bar.
+ 
+    ![click devices](./assets/devices.png )
+ 
+  + Hover over any device you want to test with and click on the Automation settings button (the gear symbol).
+ 
+    ![click auto](./assets/click_auto.png )
+ 
+  + Kobiton automatically translates your desired capabilities into code based on the options you selected. Simply copy and paste this code into your test script under the Input Capabilities section. This will replace the current configuration with the Kobiton device configuration and change the parameters in your test script.
```
Comment (thangvo38) [inline run-automation-test.md]: `Simply copy and paste this code...` What code?
Put this paragraph after you point of the desiredCaps code (next paragraph)

File: run-automation-test.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+### 3.3 Configure automation test script
+Kobiton has already provided sample scripts for automation testing, visit [here](https://github.com/kobiton/samples) for reference. 
+> Note: In this guideline, we will use the Node.js sample (samples/javascript folder) as an example.
+
+Use Kobiton’s desired capabilities settings to configure your test scripts. Our settings include all the required configurations you need to automate your test using the Appium Framework and to run your automation test on real devices in the Kobiton cloud.
+
+Open automation test script file in your repository or create a new one.
+
+Add an environment variable named `APP_URL`, replace `desiredCaps` in the script with the ones collected in the previous step. Then, replace `app` value of `desiredCaps` with the `app_url` variable.
```
Comment (thangvo38) [inline run-automation-test.md]: 1) `Add an environment variable named APP_URL` -> this doesn't sound to clear.
Suggestion: `Import the env variable APP_URL`
2) `app value of desiredCaps` -> `value of app element in desiredCaps` . Remember `app` alone  is a key/element, not a value

File: run-automation-test.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+**Attaching environment variable to Teamcity**
++ In your project in TeamCity, on the left side of the page, click **Parameters**. Then click **Add new parameter** button. 
+
+![parameters](./assets/param.png)
+
++ Add environment variables. For the **Kind** category, please choose **Environment variable (env.)**. 
+
+![env.username](./assets/param_name.png)
+
+>Note: Here are the variables we need. For the sake of this guideline, please set the name for these variables like below.
```
Comment (thangvo38) [inline run-automation-test.md]: Remove all `>` this is block because it's important information

File: run-automation-test.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+> 
+>APP_URL: You need put empty value of this variable
```
Comment (thangvo38) [inline run-automation-test.md]: Please verify if we should remove this as we discussed

File: run-automation-test.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+>
+>HOCKEYAPP_TOKEN: Hockey App API token
+>
+>KOBITON_API_KEY: Your Kobiton API Key
+>
+>KOBITON_USERNAME: Your Kobiton username
+
+![environment variables](./assets/img/2-env.png)
+
+**Follow steps below to setup TeamCity for automation testing.**
```
Comment (thangvo38) [inline run-automation-test.md]: ** Setup TeamCity for automation testing **

File: run-automation-test.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+>Note: In this example, we will be running Android App Testing. Therefore, we will use `android-app-test` script.
```
Comment (thangvo38) [inline run-automation-test.md]: You should write this before we replace our desiredCaps.
Consider these 2 suggestions: 
1) Move this part to above section
2) Add this part to above section and rewrite this as a reminder to users

File: run-automation-test.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+1. In your TeamCity project, on the left side of the page, click **Build Steps** -> **Add build step** to set build steps.
+
+![build step](./assets/build_step.png)
+
+2. On the right side of the page, for the **Runner type** category, choose **Command Line**.
+3. Add below commands in **Custom script** section.
+
+```
+  APP_URL=$(node hockeyAppApi)
```
Comment (thangvo38) [inline run-automation-test.md]: Fix this as we discussed

**Review by thangvo38 — CHANGES_REQUESTED** [review]
Review

File: README.md [hunk 10]
```diff
@@ -1,2 +1,16 @@
-# TeamCity-TestFlight-Appium-ReactNative
-Guidance on integrating Kobiton service into the mobile app build pipeline: TeamCity, TestFlight, Appium and ReactNative.
+# TeamCity-HockeyApp-Appium-ReactNative
+Guidance on integrating Kobiton service into the mobile app build pipeline: TeamCity, HockeyApp, Appium and ReactNative.
+
+If you are using:
++ TeamCity to build app
++ HockeyApp to keep all application builds
```
Comment (blackheat) [inline README.md]:
```
HockeyApp to store and distribute application builds
```

File: run-automation-test.md (hunk 3 above)
Comment (blackheat) [reply run-automation-test.md]: translates ???

File: run-automation-test.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  + On the left-hand side, you can select your preferred language, App Type, etc.
+  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section and **Hybrid/Native from Apps** in `App type` section.
+
+  ![auto](./assets/automation.png ) 
+
+### 3.3 Configure automation test script
+Kobiton has already provided sample scripts for automation testing, visit [here](https://github.com/kobiton/samples) for reference. 
+> Note: In this guideline, we will use the Node.js sample (samples/javascript folder) as an example.
+
+Use Kobiton’s desired capabilities settings to configure your test scripts. Our settings include all the required configurations you need to automate your test using the Appium Framework and to run your automation test on real devices in the Kobiton cloud.
```
Comment (blackheat) [inline run-automation-test.md]: ?? This line is confusing

File: run-automation-test.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    ![click auto](./assets/click_auto.png )
+ 
+  + Kobiton automatically translates your desired capabilities into code based on the options you selected. Simply copy and paste this code into your test script under the Input Capabilities section. This will replace the current configuration with the Kobiton device configuration and change the parameters in your test script. 
+  
+  + On the left-hand side, you can select your preferred language, App Type, etc.
+  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section and **Hybrid/Native from Apps** in `App type` section.
+
+  ![auto](./assets/automation.png ) 
+
+### 3.3 Configure automation test script
```
Comment (blackheat) [inline run-automation-test.md]: If you don't tell them to **use** the samples, at least give them a link to the tutorial on how to write test scripts :)

File: run-automation-test.md (hunk 8 above)
Comment (blackheat) [reply run-automation-test.md]: What if they use iOS ?

File: README.md [hunk 13]
```diff
@@ -1,2 +1,18 @@
...
+
+Kobiton is a mobile cloud platform that enables users to perform manual or automated testing on iOS and Android devices. 
+
+Teamcity is a Java-based build management and continuous integration server. TeamCity allows to administer build infrastructures of literally any size from a central Web interface. 
+
+HockeyApp is a service that allows developers to recruit and manage testers, distribute apps, and collect crash reports, among other things.
+
+By using TeamCity, we can automatically deploy app and run app automation test on Kobiton. 
+
+This guide will demonstrate [how to run automation tests on your app with the Kobiton service](run-automation-test.md)
```
Comment (thangvo38) [inline README.md]: Change the name. This guide demonstrate more than just `run automation test`

File: README.md [hunk 14]
```diff
@@ -1,2 +1,18 @@
-# TeamCity-TestFlight-Appium-ReactNative
-Guidance on integrating Kobiton service into the mobile app build pipeline: TeamCity, TestFlight, Appium and ReactNative.
+# TeamCity-HockeyApp-Appium-ReactNative
```
Comment (thangvo38) [inline README.md:1]: Please change the title, this doesn't look nice

File: run-automation-test.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
+# Running Automation test with Kobiton
+## Table of contents
++ [Prerequisites](#prerequisites)
++ [1. Configure TeamCity and integrate with GitHub](#1-configure-teamcity-and-integrate-with-github)
++ [2. Getting required parameters](#2-getting-required-parameters)
++ [3. Configure automation test script](#3-configure-automation-test-script)
++ [4. Run automation test on Kobiton devices](#4-run-automation-test-on-kobiton-devices)
++ [5. Fetch test session data through Kobiton REST API](#5-fetch-test-session-data-through-kobiton-rest-api)
++ [6. Feedback](#6-feedback)
+## Prerequisites
```
Comment (thangvo38) [inline run-automation-test.md]: HockeyApp account?

File: run-automation-test.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  >You can go to https://nodejs.org/en/download/ to download the latest version of node
+## 1. Configure TeamCity and integrate with GitHub
+This part will guide you to how to configure TeamCity to integrate with your automation test GitHub repository. Skip this step if you have already had TeamCity setup.
+ + For instruction on how to install and configure TeamCity server, follow [this guide](https://confluence.jetbrains.com/display/TCD18/Installing+and+Configuring+the+TeamCity+Server).
+
+Set up your project and link your app repository that we will test with Kobiton later.
++ You need to configure a connection to your GitHub repository. Instruction on how to connect GitHub with TeamCity can be found [this guide](https://confluence.jetbrains.com/display/TCD10/Integrating+TeamCity+with+VCS+Hosting+Services), in `Connecting to GitHub` section. After the connection is configured, TeamCity will be granted with `full control` of private repositories and `Write repository hooks` permission.
+ 
+>More information about TeamCity can be found in [this documentation](https://confluence.jetbrains.com/display/TCD10/TeamCity+Documentation).
+## 2. Getting required parameters
```
Comment (thangvo38) [inline run-automation-test.md]: Put a brief instruction for the whole section here would be nicer.

File: run-automation-test.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+  + Go to https://portal.kobiton.com, click **Devices** in the top navigation bar.
+ 
+    ![click devices](./assets/devices.png )
+ 
+  + Hover over any device you want to test with and click on the Automation settings button (the gear symbol).
+ 
+    ![click auto](./assets/click_auto.png )
+  
+  + On the left-hand side, you can select your preferred language, App Type, etc.
+  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section and **Hybrid/Native from Apps** in `App type` section. On the right-hand side, Kobiton automatically translates your desired capabilities into code based on the options you selected. Copy the code to prepare for the next step.
```
Comment (thangvo38) [inline run-automation-test.md]: `Kobiton automatically translates your desired capabilities into code`
Using `translates` here feels odd. Use `generates` instead

File: run-automation-test.md [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+## 3. Configure automation test script
+### 3.1. Fetching the app download URL
+To get the app download URL, we have provided a script written in NodeJS. After executing that script, you will get the latest app version, which will be used in the next step.
+
+Downloads the scipt from [this link](./getAppUrl.js) and copy to your automation test folder.
+### 3.2 Configure automation test script
+Kobiton has already provided sample scripts for automation testing, visit [here](https://github.com/kobiton/samples) for reference. 
+> Note: In this guideline, we will use the Node.js sample (samples/javascript folder) as an example. In sample/javascript folder, we will use `android-app-test` script. You can choose `android-app-test.js` if you want to run tests on Android or `ios-app-test.js` if on iOS.
+
+Use Kobiton’s desired capabilities settings to configure your test scripts. Our settings include all the required configurations you need to automate your test using the Appium Framework and to run your automation test on real devices in the Kobiton cloud.
```
Comment (thangvo38) [inline run-automation-test.md]: Fix this as we discussed

File: run-automation-test.md [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+
+Open automation test script file in your repository or create a new one.
+
+Replace `desiredCaps` in the script with the ones collected in the previous step. Then,replace value of `app` element in `desiredCaps` with the `APP_URL` environment variable (The value of `app` element is the download URL of your app. If your automation test script is in other languages, please use the proper method to access this environment variable).
```
Comment (thangvo38) [inline run-automation-test.md]: Duplicated `the value of app element`

File: run-automation-test.md [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+  deviceOrientation:  'portrait',  
+  captureScreenshots: true, 
+  app:                process.env.APP_URL, 
+  deviceGroup:        'KOBITON', 
+  deviceName:         '<DEVICE_NAME>',
+  platformVersion:    '<DEVICE_VERSION>',
+  platformName:       '<DEVICE_PLATFORM_NAME>' 
+}
+```
+### 3.3 Configure TeamCity project
```
Comment (thangvo38) [inline run-automation-test.md]: I don't see the part where we mention the user's deploying process. Please at least mention or show a picture of 2 build step to highlight that the testing step occurs after deploying step.

File: run-automation-test.md [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+2. On the right side of the page, for the **Runner type** category, choose **Command Line**.
+3. Add below commands in **Custom script** section.
+
+```
+  url=$(node getAppUrl)
+  npm install
+  APP_URL=$url npm run <YOUR_AUTOMATION_SCRIPT_EXECUTION_COMMAND>
+```
+>Note: url=$(node getAppUrl): this command is required to get the app download URL and set value of url.
```
Comment (thangvo38) [inline run-automation-test.md]: Explain `getAppUrl` is the `path to the node application`.
Reminder: This is how a directory tree in a react native project look like:
```
root
----- react native
----- testscript
```
We aren't 100% sure that the test script is in the root folder of the repo, so tell them to access the file by `path`, not just `name`

File: run-automation-test.md [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+
+  ![auto](./assets/automation.png ) 
+## 3. Configure automation test script
+### 3.1. Fetching the app download URL
+To get the app download URL, we have provided a script written in NodeJS. After executing that script, you will get the latest app version, which will be used in the next step.
+
+Downloads the scipt from [this link](./getAppUrl.js) and copy to your automation test folder.
+### 3.2 Configure automation test script
+Kobiton has already provided sample scripts for automation testing, visit [here](https://github.com/kobiton/samples) for reference. 
+> Note: In this guideline, we will use the Node.js sample (samples/javascript folder) as an example. In sample/javascript folder, we will use `android-app-test` script. You can choose `android-app-test.js` if you want to run tests on Android or `ios-app-test.js` if on iOS.
```
Comment (thangvo38) [inline run-automation-test.md]: Fix this as we discussed

**Review by thangvo38 — CHANGES_REQUESTED** [review]
Please take a look

**Pull request closed without merging**